File: Java/dfs.java
// DFS algorithm implemented in Java
//contributed by @whyanujjwhy
package org.example.algorithms;

import java.util.*;

//main class
class DFS {
  private LinkedList<Integer> adjLists[];
  private boolean visited[];

  // Graph creation
  DFS(int vertices) {
    adjLists = new LinkedList[vertices];
    visited = new boolean[vertices];

    for (int i = 0; i < vertices; i++)
      adjLists[i] = new LinkedList<Integer>();
  }

  // Adding edges src->dest
  void addEdge(int src, int dest) {
    adjLists[src].add(dest);
  }

  // DFS algorithm function
  //called recursively
  void dfsTraversal(int vertex) {
    visited[vertex] = true;
    System.out.print(vertex + " ");

    Iterator<Integer> ite = adjLists[vertex].listIterator();
    while (ite.hasNext()) {
      int adj = ite.next();
      if (!visited[adj])
        dfsTraversal(adj);
    }
  }

  public static void main(String args[]) {
    DFS g = new DFS(4); //graph of 4 nodes

    //edges in the directed graph
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 2);
    g.addEdge(2, 3);
    g.addEdge(3, 3);
    g.addEdge(2, 0);

    System.out.println("Depth First Traversal");

    g.dfsTraversal(2);  //start node is 2
  }
}

File: Kotlin/dfs.kt
class Graph(private val vertices: Int) {
    private val adjacencyList: Array<MutableList<Int>> = Array(vertices) { mutableListOf() }

    fun addEdge(u: Int, v: Int) {
        adjacencyList[u].add(v)
    }

    fun dfs(startVertex: Int) {
        val visited = BooleanArray(vertices)
        dfsRecursive(startVertex, visited)
    }

    private fun dfsRecursive(vertex: Int, visited: BooleanArray) {
        visited[vertex] = true
        print("$vertex ")

        for (adjacentVertex in adjacencyList[vertex]) {
            if (!visited[adjacentVertex]) {
                dfsRecursive(adjacentVertex, visited)
            }
        }
    }
}

fun main() {
    val graph = Graph(7)

    graph.addEdge(0, 1)
    graph.addEdge(0, 2)
    graph.addEdge(1, 3)
    graph.addEdge(1, 4)
    graph.addEdge(2, 5)
    graph.addEdge(2, 6)

    println("Depth-First Traversal (starting from vertex 0):")
    graph.dfs(0)
}
